src/decoder/stream.rs

```diff
@@ -306,6 +318,72 @@ impl EndianReader for LZWReader {
     }
 }
 
+///
+/// ## JPEG Reader (for "new-style" JPEG format (TIFF compression tag 7))
+///
+
+pub struct JpegReader {
+    buffer: io::Cursor<Vec<u8>>,
+    byte_order: ByteOrder, // for jpeg image data always ...,
+}
+// Only tested with qtables in jpegtables, but it contain both qtables and huffman tables or either
+impl JpegReader {
+    /// Constructs new JpegReader wrapping a SmartReader.
+    /// Because Jpeg compression in TIFF allows to save quantization and/or huffman tables in one
+    /// central loacation, the constructor accepts this data as `jpeg_tables` here containing either
+    /// both.
+    /// These `jpeg_tables` are simply prepended to the remaining jpeg image data.
+    /// Because these jpegtables start with a `SOI` (HEX: `0xFFD8`) or __start of image__ marker
+    /// which is also at the beginning of the remaining jpeg image data and would
+    /// confuse the jpeg renderer, one of these has to be taken off. In this case the first two
+    /// bytes of the remaining jpeg data is removed because this follows `jpeg_tables`.
+    /// Similary, `jpeg_tables` ends with a `EOI` (HEX: `0xFFD9`) or __end of image__ marker,
+    /// this has to be removed as well (last two bytes of `jpeg_tables`).
+
+    pub fn new<R>(
+        reader: &mut SmartReader<R>,
+        length: u32,
+        jpeg_tables: &Vec<u8>,
+    ) -> io::Result<JpegReader>
+    where
+        R: Read + Seek,
+    {
+        let order = reader.byte_order;
+
+        // Read jpeg image data
+        let mut segment = vec![0; length as usize];
+        reader.read_exact(&mut segment[..])?;
```

Comment:
Reading into an allocated buffer is okay for now, albeit somewhat wasteful. However, again, you should ensure the size of the allocation is limited.

Reply:
Okay, two questions:
1. what would be a less wasteful approach?
2. What do you mean with "ensuring that the allocation is limited?"

Reply:
> 1. what would be a less wasteful approach?

Wasteful refers to the fact that all data is copied into `segment`, then read into another buffer when the jpeg reader requests it–plus the clone of `jpeg_tables`. More efficient would be to avoid this intermediate buffering but it's somewhat complicated and definitely intermediate Rust. Very roughly:

* Keep the borrowed `&mut SmartReader<R>` in the struct.
* Also store the bytes of `jpeg_values` as a references (`&[u8]`).
* Complicate the impl of `Read` to first read from the values and when those are exhausted, forward to SmartReader.

But that gets tricky and its definitely not required to accept the PR.

> 2. What do you mean with "ensuring that the allocation is limited?"

The length is from `bytes[idx]` which is basically input data. We don't want to allocate 4GB of memory just because the image tells us to. You will find in the `Limits` struct some configuration for maximums. Basically you should check that the length does not exceed the `decoding_buffer_size`, although maybe even adding a third limit—named `intermediate_buffer_size` with default maybe ~128MB?—might be more accurate. This doesn't need to be 100% exact, i.e. sometimes we allocate more than once, but it should be close.

Note how the approach outlined in `1.` avoids having such a separate limit by not having an intermediate buffer.

Reply:
Thank you for the time to respond! 

> - Complicate the impl of Read to first read from the values and when those are exhausted, forward to SmartReader.

Ah this was my first idea, but I did not want to have too many moving parts, manage state and fight Rust. So I was trying to keep it simple and closely followed the approach taken in the `LZWReader`.

> ...Limits...

Ah okay I understand, will do. Although I find the size limitations overall pretty strict. E.g. I cannot load a measly 300MB TIF without having to manually adjust the limits. (But I get that limits have to be scaled up conservatively as different RAM configurations can cause problems and might require mmap)